CIPs/cip-145.md

```diff
@@ -51,17 +56,15 @@ branches are pruned. Since pruning branches that contain Data Events would resul
 a Data Event to contain multiple ancestor events in its `prev` field so that a new merge point Data Event can cover
 multiple events.
 
-1. If a stream is in a diverged state (see events `A` and `B` in fig. 5), we only consider branches that contain valid,
-   uncovered Data Events.
-2. For branches that contain valid, uncovered Data Events, we only consider the first Data Event on a branch after the
-   fork.
-3. The Data Event that is covered by the earliest Time Event wins (see event `A` in fig. 5). 
-4. If two uncovered Data Events are covered by Time Events at the same block height, the Data Event with the lower CID
-   wins.
-
-Using multi-prev Data Events allows us to reduce the number of uncovered events and converge the stream so that there is
-only a single uncovered event, without any data abandoned on pruned branches. The stream's converged/diverged state can
-be determined by looking at the `prev` fields of all the Data Events for that stream.
+1. If a stream is in a diverged state (see events `A` and `B` in fig. 5), we consider branches that contain dominant
```

Comment:
Why would time events get pruned? 
**Consider Fig. 6:** 
* Data B expires at time `3<n<4`
* Data C has a time event at `4`
* If Time 3 is pruned, then it's unclear why Data B should remain valid

Reply:
Good point. Since data A is the tip according to the rest of the rules in fig 4, I'm not sure why we should try to merge its descendants :thinking: 
![image](https://github.com/ceramicnetwork/CIPs/assets/479022/a07aa199-0c53-4c36-803f-8ae6ba769ac2)

Reply:
Yes, good question. @AaronGoldman and I had a good discussion about this scenario.
![image](https://github.com/ceramicnetwork/CIPs/assets/2236875/5de4562c-963e-4b17-83ce-0d5fc19a2af9)
In this example, `Data C` can only be created after `Data B` has been validated to either be within the expiration timeout or to already have a valid Time Event.

We can assume that the author of `Data C` validated `Data B` because `Data C` follows `Data B`.

Of course, if `Time 3` for `Data B` is available before `Data C` is created, the CID of `Time 3` will be included in `Data C`'s multi-prev.

To your point, @m0ar, including `Time 1`'s descendants allows multi-prev to cover existing history (even if pruned) 	in the stream state, whereas today, pruned events are lost from the stream state.

Reply:
> We can assume that the author of Data C validated Data B because Data C follows Data B.

Can we really make this assumption? In theory the author of `Data C` could "import" `Data B` into the log even though it's invalid, e.g. `Time 3` doesn't exist or is too late.

Reply:
> > We can assume that the author of Data C validated Data B because Data C follows Data B.
> 
> Can we really make this assumption? In theory the author of `Data C` could "import" `Data B` into the log even though it's invalid, e.g. `Time 3` doesn't exist or is too late.

Events like `Time 3` are definitely worth pinning, and potentially worth including in the multi-prev of subsequent Data Events.

We feel that the _validity_ of `Data C` does not require proof of validity of `Data A` or `Data B`. However, establishing the latter's _time bounds_ does require recording `Time 2` and `Time 3`. `Time 2` is already part of `Data C`'s history. Including `Time 3` in a future multi-prev would ensure that it also becomes part of the stream history.

Reply:
> We feel that the validity of `Data C` does not require proof of validity of `Data A` or `Data B`. However, establishing the latter's time bounds does require recording `Time 2` and `Time 3`. `Time 2` is already part of `Data C`'s history. Including `Time 3` in a future multi-prev would ensure that it also becomes part of the stream history.

Agree that `Data C`'s validity is not predicated on the validity of `A` or `B`. However, the validity of `Data B` **is** predicated on `Time 3` being available. Therefore, it doesn't seem right to say that we can prune `Time 3`.

Reply:
> > We feel that the validity of `Data C` does not require proof of validity of `Data A` or `Data B`. However, establishing the latter's time bounds does require recording `Time 2` and `Time 3`. `Time 2` is already part of `Data C`'s history. Including `Time 3` in a future multi-prev would ensure that it also becomes part of the stream history.
> 
> Agree that `Data C`'s validity is not predicated on the validity of `A` or `B`. However, the validity of `Data B` **is** predicated on `Time 3` being available. Therefore, it doesn't seem right to say that we can prune `Time 3`.

Yes, what would you think about including `Time 3` in the multi-prev of the next event to be added to the stream? That's what we meant to say here:
> Events like Time 3 are definitely worth pinning, and potentially worth including in the multi-prev of subsequent Data Events.

What if a new event `Data D` (occurring after `Time 4`) had a `prev` of `[Time 4, Time 3]`? Even though `Time 3` will not take precedence during tip selection, it will always remain part of the stream history.

Or, let's say, `Time 3` didn't show up until we already had `Time 4` -> `Data D` -> `Time 5` -> `Data E` -> `Time 6`. Then `Data F` (occurring after `Time 6`) would have a `prev` of `[Time 6, Time 3]`. `Data B` would remain in an "unverified" state until `Time 3` was discovered.

Tracking `Data B`'s validity this way would be a little more complicated than the usual flow, but always possible. Moreover, now _all_ events related to the stream would be part of the DAG, which is great.

Reply:
Yes, this is what I mean! Just wanted to be clear that `Time 3` can't be pruned if we expect the creator of `Data D` to include it in *prev*.

Reply:
@smrz2001 since it seems hat we are in agreement, maybe you can update the language to be clear that the TimeEvent doesn't get pruned?